File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/communication/RequestSubmitter.java
package lunalauf.rms.utilities.network.communication;

import com.google.gson.JsonParseException;
import javafx.concurrent.Task;
import lunalauf.rms.utilities.network.client.Client;
import lunalauf.rms.utilities.network.communication.message.Message;
import lunalauf.rms.utilities.network.communication.message.request.Request;
import lunalauf.rms.utilities.network.communication.message.response.Response;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class RequestSubmitter {

    private final Consumer<Response> responseHandler;
    private final Consumer<ErrorType> errorHandler;
    private final ExecutorService executor;

    private final int numMessagesToDump = 5;

    public RequestSubmitter(Consumer<Response> responseHandler, Consumer<ErrorType> errorHandler) {
        this.responseHandler = responseHandler;
        this.errorHandler = errorHandler;
        this.executor = Executors.newSingleThreadExecutor();
    }

    public void submit(Request request, Client client) {
        SubmissionTask task = new SubmissionTask(request, client);

        task.setOnSucceeded(event -> {
            try {
                Response response = task.get();
                responseHandler.accept(response);
            } catch (Exception e) {
                errorHandler.accept(ErrorType.UNWANTED_TERMINATION);
            }
        });
        task.setOnCancelled(event -> errorHandler.accept(ErrorType.UNWANTED_TERMINATION));
        task.setOnFailed(event -> {
            Throwable exception = task.getException();
            ErrorType error;
            if (exception instanceof Client.NoAnswerException)
                error = ErrorType.RESPONSE_TIMEOUT;
            else if (exception instanceof Client.NoConnectionException)
                error = ErrorType.DISCONNECTED;
            else if (exception instanceof CorruptedMessageException)
                error = ErrorType.CORRUPTED_SERVER_MESSAGE;
            else
                error = ErrorType.CORRUPTED_SERVER_MESSAGE;

            errorHandler.accept(error);
        });

        executor.execute(task);
    }

    public void shutdown() {
        try {
            executor.shutdown();
            if (!executor.awaitTermination(3, TimeUnit.SECONDS))
                executor.shutdownNow();
        } catch (InterruptedException ignored) {
        }
    }

    private class SubmissionTask extends Task<Response> {
        private final Request request;
        private final Client client;

        SubmissionTask(Request request, Client client) {
            this.request = request;
            this.client = client;
        }

        @Override
        protected Response call() throws Client.NoAnswerException, Client.NoConnectionException, CorruptedMessageException {
            if (client == null || !client.isConnected())
                throw new Client.NoConnectionException();

            client.send(MessageProcessor.toJsonString(request));
            Response response = getMatchingResponse(request);
            if (response == null)
                throw new CorruptedMessageException();
            return response;
        }

        private Response getMatchingResponse(Request request) throws Client.NoAnswerException, Client.NoConnectionException, CorruptedMessageException {
            int it = 0;
            while (it < numMessagesToDump) {
                String receivedString = client.receive();
                try {
                    Message receivedMessage = MessageProcessor.fromJsonString(receivedString);
                    if (receivedMessage instanceof Response response) {
                        if (response.messageId == request.messageId || response.messageId == -1)
                            return response;
                    }
                } catch (JsonParseException e) {
                    throw new CorruptedMessageException(e);
                }
                it++;
            }
            throw new Client.NoAnswerException();
        }
    }

    public static class CorruptedMessageException extends Exception {
        public CorruptedMessageException() {
            super();
        }

        public CorruptedMessageException(Throwable cause) {
            super(cause);
        }
    }

}

File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/communication/RequestSubmitter.kt
package lunalauf.rms.utilities.network.communication

import com.google.gson.JsonParseException
import kotlinx.coroutines.*
import lunalauf.rms.utilities.network.client.Connection
import lunalauf.rms.utilities.network.communication.message.request.Request
import lunalauf.rms.utilities.network.communication.message.response.Response
import org.slf4j.LoggerFactory

class RequestSubmitter(
    private val responseHandler: (Response) -> Unit,
    private val errorHandler: (ErrorType) -> Unit
) {
    companion object {
        private const val NUM_MESSAGES_TO_DUMP = 5
    }

    private val logger = LoggerFactory.getLogger(this::class.java)

    private val scope = CoroutineScope(Dispatchers.IO)
    private var job: Job? = null

    @Synchronized
    fun submit(
        request: Request,
        connection: Connection
    ) {
        job = scope.launch {
            try {
                val response = run(request, connection)
                responseHandler(response)
            } catch (e: Throwable) {
                val error = when (e) {
                    is CancellationException -> ErrorType.UNWANTED_TERMINATION
                    is Connection.NoAnswerException -> ErrorType.RESPONSE_TIMEOUT
                    is Connection.NoConnectionException -> ErrorType.DISCONNECTED
                    is CorruptedMessageException -> ErrorType.CORRUPTED_SERVER_MESSAGE
                    else -> ErrorType.CORRUPTED_SERVER_MESSAGE
                }
                errorHandler(error)
            }
        }
    }

    @Synchronized
    fun shutdown() {
        job?.cancel()
    }

    @Throws(
        Connection.NoAnswerException::class,
        Connection.NoConnectionException::class,
        CorruptedMessageException::class,
        CancellationException::class
    )
    private fun CoroutineScope.run(
        request: Request,
        connection: Connection
    ): Response {
        connection.sendMessage(request)
        logger.info("Request sent: {}", request)

        var it = 0
        while (it < NUM_MESSAGES_TO_DUMP) {
            if (!isActive)
                throw CancellationException()

            try {
                val receivedMessage = connection.receiveMessage()
                if (receivedMessage is Response) {
                    if (receivedMessage.messageId == request.messageId || receivedMessage.messageId == -1L) {
                        logger.info("Received message: {}", receivedMessage)
                        return receivedMessage
                    }
                }
            } catch (e: JsonParseException) {
                logger.warn("Received corrupted server message", e)
                throw CorruptedMessageException(e)
            }
            it++
        }
        logger.error("Server did not respond")
        throw Connection.NoAnswerException()
    }

    class CorruptedMessageException : Exception {
        constructor() : super()
        constructor(cause: Throwable?) : super(cause)
    }
}
